packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -237,126 +169,170 @@ fn eval_file<'s>(scope: &mut v8::HandleScope<'s>, path: &str) -> Result<Value<'s
     let js_source_code = new_js_string(scope, &source_code);
     let mut try_catch = v8::TryCatch::new(scope);
     let script = v8::Script::compile(&mut try_catch, js_source_code, None).ok_or_else(|| {
-        let exception = try_catch.exception().unwrap();
-        let exception_string = exception
-            .to_string(&mut try_catch)
-            .unwrap()
-            .to_rust_string_lossy(&mut try_catch);
+        let exception = read_try_catch_exception(&mut try_catch).unwrap();
 
-        Error::Eval(path.into(), format!("Compile error: {exception_string}"))
+        Error::Eval(path.into(), format!("Compile error: {exception}"))
     })?;
 
     script.run(&mut try_catch).ok_or_else(|| {
-        let exception = try_catch.exception().unwrap();
-        let exception_string = exception
-            .to_string(&mut try_catch)
-            .unwrap()
-            .to_rust_string_lossy(&mut try_catch);
+        let exception = read_try_catch_exception(&mut try_catch).unwrap();
 
-        Error::Eval(path.into(), format!("Execution error: {exception_string}"))
+        Error::Eval(path.into(), format!("Execution error: {exception}"))
     })
 }
 
-fn import_file<'s>(
-    scope: &mut v8::HandleScope<'s>,
-    context: v8::Local<'s, v8::Context>,
-    path: &str,
-    args: &[Value<'s>],
-) -> Result<Value<'s>> {
-    let v = eval_file(scope, path)?;
-    let f: Function<'_> = v.try_into().map_err(|err| {
-        Error::FileImport(
-            path.into(),
-            format!("Could not convert file into Function: {err}"),
-        )
-    })?;
+/// Cache module to not evaluate them twice
+struct ModuleMap {
+    path_to_module: HashMap<String, v8::Global<v8::Module>>,
+}
+
+impl ModuleMap {
+    fn new() -> Self {
+        Self {
+            path_to_module: HashMap::new(),
+        }
+    }
+
+    fn import_module<'s>(
```

Comment:
Having this function on `ModuleMap` feels odd. The map does not import the module (thus you don't need a `self` parameter). To me `ModuleMap` only seems to be an optimization/implementation detail for this function.

Reply:
Do I see correctly, that the `ModuleMap` is a cache to map a module to a given path? My point is that the logic for importing a module shouldn't be in a struct that is supposed to cache the module.

Reply:
Yes it caches modules, but that's not just an optimization.
At first I didn't want this to be part of this PR at all. But the JS spec seems to rely on modules only being evaluated once.

> Each time this operation is called with a specific referencingScriptOrModule, specifier pair as arguments it must return the same Module Record instance if it completes normally.

> After reviewing the specification, you know that a JavaScript module is evaluated once. Also, when importing modules from the same path, the same module instance is returned.

[source](https://dmitripavlutin.com/javascript-module-import-twice/)

So to me we have this struct responsible of handling modules. And adding modules is one of its responsibility.

Reply:
Good quote, can you add that as documentation?

> So to me we have this struct responsible of handling modules. And adding modules is one of its responsibility.

The `ModuleMap` is stored *inside* the scope, so why it is responsible for importing the module? The scope *uses* the module map to satisfy the expectations from V8. In my head, we use the scope like this:
```rust
struct Scope {
    module_map: ModuleMap,
    v8_stuff: ...,
}

impl Scope {
    fn load_module(&mut self, module_path: ...) -> ... {
        if !self.module_map.contains(module_path) {
            self.import_module(module_path);
        }
        ...
    }
}
```

Reply:
To me the fact that the `ModuleMap` is stored inside the scope is the implementation detail. It could very much not be but it's convenient.
This way we are 100% sure it's in the same thread as V8 and it'll likely drop its references when the scope is dropped.

Added quotes in https://github.com/hashintel/hash/pull/493/commits/48de24784315d7215ddca7ce8a3c5aa09017906a

Reply:
`lazy_static` requires the type to be `Sync`, same applies to `Mutex` etc. so there is really no sound way to share the scope between threads as long as it's `!Sync`. Let's stay tuned if multithreading gets added at some point. 🙂

Reply:
I was referring to `ModuleMap` in a `lazy_static`. But `v8::Global` is `!Send + !Sync` too so never mind.